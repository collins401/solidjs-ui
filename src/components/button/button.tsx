import { createEffect, createMemo, createSignal, JSX, Show, splitProps } from 'solid-js';
import { Loading } from '../loading';
import { isPromise } from '../utils';
import { cva, VariantProps } from 'class-variance-authority';
import clsx from 'clsx';

export type ButtonBaseProps = VariantProps<typeof button>;
const button = cva('so-button', {
  variants: {
    type: {
      primary: ['bg-primary', 'text-white', 'border-primary'],
      secondary: ['bg-secondary', 'text-white', 'border-secondary'],
      default: ['border border-button']
    },
    size: {
      small: ['text-12px', 'h-7', 'px-2 rounded'],
      medium: ['text-14px h-8 px-3 rounded'],
      large: ['text-16px', 'h-12', 'px-4 rounded-lg font-medium']
    },
    /** outline */
    outline: {
      true: ['!bg-transparent border border-current']
    }
  },
  compoundVariants: [
    { type: 'primary', outline: true, class: '!text-primary' },
    { type: 'secondary', outline: true, class: '!text-secondary' }
  ],
  defaultVariants: {
    type: 'default',
    size: 'medium'
  }
});

export interface ButtonProps
  extends ButtonBaseProps,
    Omit<JSX.ButtonHTMLAttributes<HTMLButtonElement>, 'type'> {
  style?: any;
  class?: any;
  classList?: any;
  loading?: boolean | 'auto';
  block?: boolean;
  children?: string | JSX.Element;
  onClick?: (val: any) => void;
}
export function Button(props: ButtonProps) {
  const [innerLoading, setInnerLoading] = createSignal(false);
  const [local, rest] = splitProps(props, [
    'children',
    'type',
    'size',
    'outline',
    'class',
    'loading',
    'onClick'
  ]);

  createEffect(() => {
    if (local.loading && local.loading !== 'auto') {
      setInnerLoading(local.loading);
    } else {
      setInnerLoading(false);
    }
  });

  async function handleClick(e) {
    if (!local.onClick || innerLoading()) return;
    if (isPromise(local.onClick)) {
      try {
        if (local.loading === 'auto') {
          setInnerLoading(true);
          await local.onClick(e);
          setInnerLoading(false);
        }
      } catch (e) {
        setInnerLoading(false);
        throw e;
      }
    } else {
      local.onClick(e);
    }
  }

  return (
    <button
      onClick={handleClick}
      class={clsx(
        button({
          type: local.type,
          size: local.size,
          outline: local.outline,
          class: local.class
        }),
        innerLoading() ? 'pointer-events-none' : ''
      )}
      {...rest}
    >
      <Show when={typeof local.children !== 'string' || innerLoading()} fallback={local.children}>
        <div class="flex items-center justify-center space-x-1">
          <Show when={innerLoading()}>
            <Loading />
          </Show>
          <span>{local.children}</span>
        </div>
      </Show>
    </button>
  );
}
