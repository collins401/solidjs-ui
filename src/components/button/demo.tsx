import { Button } from './index';
export default function InputDemo() {
  return (
    <div class="py-3">
      <div class="text-color/60 mb-1 mx-5">Basic</div>
      <div class="bg-color flex flex-wrap py-3 px-5 gap-1">
        <Button type="primary">主题按钮</Button>
        <Button type="secondary">次要按钮</Button>

        <Button>默认按钮</Button>

        <Button class="bg-blue-500 text-white">自定义颜色</Button>
      </div>
      <div class="text-color/60 mb-1 mx-5">Outline</div>
      <div class="bg-color flex flex-wrap py-3 px-5 gap-1">
        <Button outline type="primary">
          主题线框
        </Button>
        <Button outline type="secondary">
          次要线框
        </Button>
        <Button outline class="text-amber-500">
          自定义线框
        </Button>
      </div>
      <div class="text-color/60 mb-1 mx-5">Light</div>
      <div class="bg-color flex flex-wrap py-3 px-5 gap-1">
        <Button type="primary" class="!bg-primary/20 !text-primary">
          主题背景
        </Button>
        <Button type="secondary" class="!bg-secondary/20 !text-secondary">
          次要背景
        </Button>
        <Button type="primary" class="!bg-blue-500/20 !text-blue-500">
          蓝色背景
        </Button>
        <Button type="primary" class="!bg-purple-500/20 !text-purple-500">
          紫色背景
        </Button>
      </div>
      <div class="text-color/60 mt-3 mb-1 mx-5">Disabled</div>
      <div class="bg-color py-3 px-5 space-x-[3px]">
        <Button disabled type="primary">
          主题禁用
        </Button>
        <Button disabled type="secondary">
          secondary禁用
        </Button>
        <Button disabled>默认禁用</Button>
      </div>
      <div class="text-color/60 mb-1 mx-5 mt-5">Custom size style rounded</div>
      <div class="bg-color p-5">
        <div class="space-x-1">
          <Button type="primary" size="small">
            小按钮
          </Button>
          <Button type="primary">默认按钮</Button>
          <Button type="primary" size="large">
            大按钮
          </Button>
          <Button class="bg-yellow-600 text-white  !h-14 rounded-lg text-lg">
            自定义大小 h-14
          </Button>
        </div>
        <div class="mt-2 space-y-[5px]">
          <Button type="primary" class="text-lg rounded-full !h-16 w-full">
            !h-16 w-full rounded-full text-lg
          </Button>
        </div>
      </div>
      <div class="text-color/60 mb-1 mx-5 mt-5">Loading</div>
      <div class="bg-color p-5 space-x-[5px]">
        <Button type="primary" loading />
        <Button type="primary" loading disabled>
          disabled
        </Button>
        <Button type="secondary" loading>
          加载中...
        </Button>
      </div>
      <div class="text-color/60 mb-1 mx-5 mt-5">Debounce</div>
      <div class="bg-color p-5 space-x-[5px]">
        <Button
          type="primary"
          loading="debounce"
          onClick={async () => {
            console.log('click');
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve(true);
              }, 3000);
            });
          }}
        >
          异步防抖
        </Button>
      </div>
      <div class="text-color/60 mb-1 mx-5 mt-5">Beautiful</div>
      <div class="bg-color p-5 space-x-[5px]">
        <Button class="!border-[red] border-dashed text-[red]" plain>
          虚线边框
        </Button>
        <Button type="primary" class="bg-gradient-to-r from-[#ff0000] to-[#7367F0] text-white">
          渐变按钮
        </Button>
      </div>
    </div>
  );
}
