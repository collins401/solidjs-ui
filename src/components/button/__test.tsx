import { describe, expect, test } from 'vitest';
import { fireEvent, render, screen } from 'solid-testing-library';
import { Button } from './button';

describe('<Button />', () => {
  test('renders', () => {
    const { container, unmount } = render(() => <Button type="primary">测试Button</Button>);
    expect(container.innerHTML).toMatchSnapshot();
    unmount();
  });

  test('updates', async () => {
    const { container, unmount, queryByText } = render(() => <Button>测试Button</Button>);
    const button = queryByText('测试Button');
    const buttonClicked = new Promise((resolve) => {
      const handler = (ev) => {
        button.removeEventListener('click', handler);
        resolve(ev);
      };
      button.addEventListener('click', handler);
    });
    fireEvent.click(button);
    await buttonClicked;
    expect(container.innerHTML).toMatchSnapshot();
    unmount();
  });
});
