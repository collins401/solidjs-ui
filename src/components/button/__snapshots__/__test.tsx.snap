// Vitest Snapshot v1

exports[`<Button /> > renders 1`] = `"<button class=\\"go3912777184  go210439308\\" type=\\"button\\"><div class=\\"flex-center space-x-[3px]\\">测试Button</div></button>"`;

exports[`<Button /> > updates 1`] = `"<button class=\\"go3912777184  go3533386438\\" type=\\"button\\"><div class=\\"flex-center space-x-[3px]\\">测试Button</div></button>"`;
